[docs] Fix MongoDB connector docs that incorrectly state it always streams from the primary (#7534)
debezium/debezium: documentation/modules/ROOT/pages/connectors/mongodb.adoc

In this case, after a restart the connector detects the missing oplog operations, performs a snapshot, and then proceeds to stream changes.

The MongoDB connector is also quite tolerant of changes in membership and leadership of the replica sets, of additions or removals of shards within a sharded cluster, and network problems that might cause communication failures.

> By default, the connector streams changes from the replica set's primary node, but you can configure it to stream from a secondary by setting the read preference in the connection string (see xref:read-preference[]).

Reviewer: ```suggestion
By default, the connector streams changes from the replica set's primary node, but you can configure it to stream from a secondary by setting the xref:read-preference[read preference] in the connection string.
```

> If the member that the connector reads from becomes unavailable -- for example, because the replica set elects a new primary -- the connector stops streaming and automatically reconnects to another suitable member based on the configured read preference.

Reviewer: ```suggestion
If the replica set member that the connector reads from becomes unavailable -- for example, because the replica set elects a new primary -- the connector stops streaming and automatically reconnects to another suitable member based on the configured read preference.
```

> If the member that the connector is reading from becomes unavailable -- for example, because the replica set elects a new primary -- the connector stops streaming and automatically reconnects to another suitable member according to the configured read preference.

Reviewer: ```suggestion
If the member that the connector reads from becomes unavailable -- for example, because the replica set elects a new primary -- the connector stops streaming and automatically reconnects to another suitable member based on the configured read preference.
```

> Similarly, if the connector is unable to communicate with the replica set, it attempts to reconnect (using exponential backoff so as to not overwhelm the network or replica set).

Reviewer: ```suggestion
Similarly, if the connector is unable to communicate with the replica set, it attempts to reconnect by using exponential backoff so as to not overwhelm the network or replica set.
```

[id="read-preference"]
=== Read Preference

The connector captures changes by using MongoDB change streams, which can be served by any member of a replica set.

> You control which member the connector reads from by setting the standard MongoDB link:https://www.mongodb.com/docs/manual/core/read-preference/[`readPreference`] parameter in the xref:mongodb-property-mongodb-connection-string[`mongodb.connection.string`].

Reviewer: ```suggestion
You specify the replica member that the connector reads from by configuring the MongoDB link:https://www.mongodb.com/docs/manual/core/read-preference/[`readPreference`] parameter in the xref:mongodb-property-mongodb-connection-string[`mongodb.connection.string`].
```

> If you do not specify a read preference, the driver default of `primary` applies, so the connector streams changes from the primary node.

Reviewer: ```suggestion
If you do not specify a read preference, the driver default of `primary` applies, and the connector streams changes from the primary node.
```

Reviewer 2: @dutta1kartik3 this appears to still be open.

> Reading from a secondary (for example, `readPreference=secondaryPreferred`) can reduce load on the primary.

Reviewer: ```suggestion
Configuring the connector to read from a secondary node, for example, by setting `readPreference=secondaryPreferred`, can reduce the load on the primary.
```

Reviewer 2: @dutta1kartik3 this appears to still be open.

> Be aware, however, that change streams emit only majority-committed events, and a secondary can lag behind the primary by its replication delay, so events captured from a secondary can arrive later than they would from the primary.

Reviewer: ```suggestion
However, because change streams emit only majority-committed events, be aware that replication delay results in secondary nodes lagging behind the primary.
As a result, events captured from a secondary can arrive later than they would from the primary.
```

[id="read-preference"]
=== Read Preference

> The connector captures changes by using MongoDB change streams, which can be served by any member of a replica set.

Reviewer: ```suggestion
The connector uses MongoDB change streams to capture changes.
Because the change streams API can read data from secondary nodes, you can configure the connector to capture changes from any member of the replica set.
```

You specify the replica member that the connector reads from by configuring the MongoDB link:https://www.mongodb.com/docs/manual/core/read-preference/[`readPreference`] parameter in the xref:mongodb-property-mongodb-connection-string[`mongodb.connection.string`].
If you do not specify a read preference, the driver default of `primary` applies, and the connector streams changes from the primary node.

> Configuring the connector to read from a secondary node, for example, by setting `readPreference=secondaryPreferred`, can reduce the load on the primary.

Reviewer: ```suggestion
To reduce the load on the primary, it's possible to configure the connector to read from a secondary node, by setting `readPreference=secondaryPreferred`.
```

However, because change streams emit only majority-committed events, be aware that replication delay results in secondary nodes lagging behind the primary.

> As a result, events captured from a secondary can arrive later than they would from the primary.

Reviewer: ```suggestion
Under normal operating conditions, configuring the connector to read from a secondary node does not significantly increase data-capture latency. 
However, events captured from a secondary can arrive later than they would from the primary due to resource contention or network latency.
Additional latency can also occur because change streams emit events only after changes are majority committed. 
A change is majority committed when a majority of voting replica set members confirm that they have written the change to their journals. 
As a result, a secondary node can emit an event only after it receives confirmation that the change has been majority committed.
```

[[mongodb-tailing-the-oplog]]

After the connector task for a replica set records an offset, it uses the offset to determine the position in the oplog where it should start streaming changes.

> The task then opens a MongoDB change stream -- against the deployment, a database, or a collection, depending on the xref:mongodb-property-capture-scope[`capture.scope`] setting -- and starts streaming changes from that position.

Reviewer: ```suggestion
The task then opens a MongoDB change stream and starts streaming changes from the offset position.
Depending on the value of the xref:mongodb-property-capture-scope[`capture.scope`] setting, the task streams changes from the deployment, a specific database, or a collection.
```

> The change stream is served by the replica set member that matches the configured read preference (the primary, by default); see xref:read-preference[].

Reviewer: ```suggestion
The change stream is served by the replica set member that matches the configured read preference (the primary, by default).
For more information, see xref:read-preference[read-preference].
```
Because the ModuleID for this topic is set to a string that differs from the original anchor ID, you must specify the link text so that the link renders properly in downstream editions.

To avoid unexpected results, consumers must be able to handle duplicate messages.
====

> As described in xref:read-preference[], the connector tasks stream changes from the replica set member that matches the configured read preference.

Reviewer: ```suggestion
As described in xref:read-preference[read preferences], the connector streams changes from the replica set member that is specified by the `readPreference` parameter in the MongoDB connection string.
```

> By default this is the primary, which lets the connector capture operations with the lowest latency; you can configure the connector to read from a secondary instead, at the cost of additional latency from replication lag.

Reviewer: ```suggestion
By default, the connector captures from the primary node, which result in capture operations with the lowest latency.
Optionally, you can configure the connector to read from a secondary, but capturing from a secondary typically results in additional latency due to replication lag.
```

To avoid unexpected results, consumers must be able to handle duplicate messages.
====

As described in xref:read-preference[read preferences], the connector streams changes from the replica set member that is specified by the `readPreference` parameter in the MongoDB connection string.
By default, the connector captures from the primary node, which result in capture operations with the lowest latency.

> Optionally, you can configure the connector to read from a secondary, but capturing from a secondary typically results in additional latency due to replication lag.

Reviewer: ```suggestion
Optionally, you can configure the connector to read from a secondary, but capturing from a secondary can introduce additional latency due to replication lag.
```

To avoid unexpected results, consumers must be able to handle duplicate messages.
====

As described in xref:read-preference[read preferences], the connector streams changes from the replica set member that is specified by the `readPreference` parameter in the MongoDB connection string.
By default, the connector captures from the primary node, which result in capture operations with the lowest latency.

> Optionally, you can configure the connector to read from a secondary, but capturing from a secondary can introduce additional latency due to replication lag.

Reviewer: ```suggestion
Optionally, you can configure the connector to read from a secondary, but capturing from a secondary can introduce additional latency due to replication lag.
For more information, see xref:read-preference[Read preferences].
```

> If the member that the connector is reading from becomes unavailable -- for example, when the replica set elects a new primary -- the connector stops streaming changes, reconnects to another suitable member, and resumes streaming from the same position.

Reviewer: ```suggestion
If the replica set member that the connector reads from becomes unavailable, the connector stops streaming changes, reconnects to another suitable member, and resumes streaming from the same position.
```
SInce we already describe this earlier in the document, I think that it's ok to drop the detail here.

[id="mongodb-becomes-unavailable-while-debezium-is-running"]
=== MongoDB becomes unavailable

> Once the connector is running, if the replica set member that the connector reads from (the primary, by default) becomes unavailable or unreachable, the connector repeatedly attempts to reconnect to a suitable member, using exponential backoff to prevent saturating the network or servers. If no suitable member is available after the configurable number of connection attempts, the connector fails.

Reviewer: ```suggestion
If the connector is running and the replica set member that it reads from becomes unavailable, it keeps trying to reconnect to another suitable member, using exponential backoff to avoid overloading the network or servers.
If the connector is unable to connect to a replica set member after the configured number of connection attempts, the connector fails.
```
